exploit/main.js

```diff
@@ -12,202 +12,211 @@ var config = require('../config');
+            var words = Object.keys(reservedWords.KEYWORDS['6-strict']);
+            var code = data.args.join(' ');
+            var ret = true;
+            if (~code.indexOf('window.response')) {
+                ret = false;
+            }
+            for (var i = 0; i < words.length; i++) {
+                var w = words[i];
+                var s = code.substr(0, w.length);
+                if (s === w) {
+                    ret = false;
+                }
+            }
+            if (ret) {
+                code = 'window.response = ' + code;
+            }
+            window.response = null;
+            eval('with (sc) { ' + code + '}');
+            socket.send(JSON.stringify({
+                type: 'evald',
+                response: window.response || 'no output'
+            }));
+        } else if (data.cmd === 'evalfile') {
+            var code = data.args[0];
+            eval('with (sc) {\n' + code + '\n}');
+            socket.send(JSON.stringify({
+                type: 'evald',
+                response: 'no output'
+            }));
+        } else if (data.cmd === 'reboot') {
+            socket.send(JSON.stringify({
+                type: 'rebooting',
+                response: 'Rebooting...'
+            }));
+            sc.ipcMsg(1).sendTo("bpc").assertOk();
+        } else if (data.cmd === 'runnro') {
+            runNro(data.args[0], data.args.slice(1));
+            socket.send(JSON.stringify({
+                type: 'rannro',
+                response: 'no output'
+            }));
+        }
+    };
 }
 
 function setupListener (sc) {
-	socket = new WebSocket('ws://' + window.location.hostname + ':8100');
-
-	var handlerFcn = handler(sc, socket);
-	socket.onmessage = (evt) => {
-		try {
-			handlerFcn(evt);
-		} catch (e) {
-			window.onerror(e.message, null, e.line, null, e);
-		}
-	};
-
-	socket.onopen = function () {
-		sc.getService("set:cal", (setcal) => {
-			socket.send(JSON.stringify({
-				type: "identification",
-				mac: sc.ipcMsg(6).sendTo(setcal).assertOk().data,
-				version: sc.version
-			}));
-		});
-		utils.log("Connected to PC...");
-	};
-
-	socket.onerror = function() {
-		utils.log("socket error, attempting to reconnect in 5 seconds...");
-		window.setTimeout(() => {
-			socket.close();
-			setupListener(sc);
-		}, 5000);
-	};
-	
-	socket.onclose = function() {
-		utils.log("socket closed, attempting to reconnect in 5 seconds...");
-		window.setTimeout(() => {
-			utils.log("attempting to reconnect...");
-			setupListener(sc);
-		}, 5000);
-	};
+    socket = new WebSocket('ws://' + window.location.hostname + ':8100');
+
+    var handlerFcn = handler(sc, socket);
+    socket.onmessage = (evt) => {
+        try {
+            handlerFcn(evt);
+        } catch (e) {
+            window.onerror(e.message, null, e.line, null, e);
+        }
+    };
+
+    socket.onopen = function () {
+        /* Only use set:cal if we have smhax. */
+        if (utils.smhax.get()) {
```

Comment:
Can we replace this with some console-unique string that comes from somewhere the browser has in NPDM?